File: Lab4_22/4.cpp
// Метод прогонки
#include <iostream>
#include <vector>

using namespace std;

vector<double> Metod_Progonki(vector<vector<double>>& A, vector<double>& B) {
    // Известные константы
    double k1 = -A[0][1];
    double m1 = B[0];
    double k2 = -A[A.size() - 1][A[0].size() - 2];
    double m2 = B[B.size() - 1];
    
    double alfa = k1;
    double beta = m1;
    
    // Поиск alfa и beta
    int c = 2;
    int a = 0;
    int b = 1;
    vector<double> alf{alfa};
    vector<double> bet{beta};
    
    for (int i = 1; i < A.size() - 1; ++i) {
        if (c < A[i].size()) {
            beta = (B[i] - A[i][a] * beta) / (A[i][a] * alfa + A[i][b]);
            alfa = -A[i][c] / (A[i][a] * alfa + A[i][b]);
            a++;
            b++;
            c++;
            alf.push_back(alfa);
            bet.push_back(beta);
        } else {
            break;
        }
    }
    
    // Расчет y
    double y = (k2 * beta + m2) / (1 - k2 * alfa);
    vector<double> otv{y};
    
    // Цикл для обратного вычисления ответа
    for (int i = alf.size() - 2; i >= 0; --i) {
        y = alf[i] * y + bet[i];
        otv.push_back(y);
    }
    
    return otv;
}

int main() {
    vector<vector<double>> A = {{-14, 6, 0}, {2, 7, 0}, {-7, -18, -9}, {2, -13, 2}, {-7, -7, 0}};
    vector<double> B = {82, -51, -46, 111, 35};

    // Вызов функции и вывод на экран
    for (auto& row: A) {
        for (auto& num: row) {
            cout << num << " ";
        }
        cout << endl;
    }
    for (double num : B) {
        cout << num << " ";
    }
    cout << endl;

    vector<double> Result = Metod_Progonki(A, B);
    cout << "Результат: ";
    for (double num : Result) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: Lab4_22/4.py
# Метод прогонки
import numpy as np

def Metod_Progonki(A, B):
    # Известные константы
    k1 = -A[0][1]
    m1 = B[0]
    k2 = -A[-1][-2]
    m2 = B[-1]
    
    alfa = k1
    beta = m1
    
    # Поиск alfa и beta
    c = 2
    a = 0
    b = 1
    alf = [alfa]
    bet = [beta]
    
    for i in range(1, A.shape[0] - 1):
        if c < A.shape[1]:
            beta = (B[i] - A[i][a] * beta) / (A[i][a] * alfa + A[i][b])
            alfa = -A[i][c] / (A[i][a] * alfa + A[i][b])
            a += 1
            b += 1
            c += 1
            alf.append(alfa)
            bet.append(beta)
        else:
            break
    
    # Расчет y
    y = (k2 * beta + m2) / (1 - k2 * alfa)
    otv = [y]
    
    # Цикл для обратного вычисления ответа
    for i in range(len(alf) - 1, -1, -1):
        y = alf[i] * y + bet[i]
        otv.append(y)
    
    # Переворачиваем значения в списке
    otvet = []
    for i in reversed(otv):
        otvet.append(i)
    
    return otvet

# Задаем матрицу системы 5x3
A = np.array([[-14, 6, 0], [2, 7, 0], [-7, -18, -9], [2, -13, 2], [-7, -7, 0]])
B = np.array([82, -51, -46, 111, 35])

# Вызов функции и вывод на экран
print(A)
print(B)
Result = Metod_Progonki(A, B)
print(f'Результат: {Result}')

